jwt/keyset.go

```diff
@@ -0,0 +1,239 @@
+package jwt
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"strings"
+
+	"github.com/coreos/go-oidc"
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// KeySet represents a set of keys that can be used to verify the signatures of JWTs.
+// A KeySet is expected to be backed by a set of local or remote keys.
+type KeySet interface {
+
+	// VerifySignature parses the given JWT, verifies its signature, and returns the claims in its payload.
+	// The given JWT must be of the JWS compact serialization form.
+	VerifySignature(ctx context.Context, token string) (claims map[string]interface{}, err error)
+}
+
+// jsonWebKeySet verifies JWT signatures using keys obtained from a JWKS URL.
+type jsonWebKeySet struct {
+	remoteJWKS oidc.KeySet
+}
+
+// staticKeySet verifies JWT signatures using local public keys.
+type staticKeySet struct {
+	publicKeys []crypto.PublicKey
+}
+
+// NewOIDCDiscoveryKeySet returns a KeySet that verifies JWT signatures using keys from the
+// JSON Web Key Set (JWKS) published in the discovery document at the given discoveryURL.
```

Comment:
Hum... if we're going to use discoveryURL as a name, then we should specify that ".well-known/openid-configuration" shouldn't be included.  We could also change the name to "issuer", which explicity doesn't include the URL fragment.

Reply:
I agree. I think changing to "issuer" makes more sense here. Done in decf7c6.